File: src/shared/domain/value_object/contact/phone_number_value_object.py
import re
from ..invalid_argument_exception import InvalidArgumentException
from .string_value_object import StringValueObject


class PhoneNumberValueObject(StringValueObject):
    def __init__(self, value: str):
        cleaned_value = re.sub(r"[\s\-\(\)]", "", value)
        super().__init__(cleaned_value)
        self._ensure_is_valid_phone()

    def _ensure_is_valid_phone(self) -> None:
        pattern = r"^\+?[1-9]\d{1,14}$"
        if not re.match(pattern, self._value):
            raise InvalidArgumentException(
                field="phone_number",
                value=self._value,
                message="Invalid phone number format",
            )

File: src/core/context/domain/value-objects/contact/phone-number-value-object.ts
import { StringValueObject } from '../primitives/string-value-object'
import { InvalidValueObjectError } from '../errors/invalid-value-object-error'

/**
 * Represents an international phone number
 */
export class PhoneNumberValueObject extends StringValueObject {
	private static readonly PHONE_REGEX = /^\+(?:[0-9]\x20?){6,14}[0-9]$/

	constructor(value: string) {
		super(value.replace(/[\s-]/g, '')) // Normaliza el número
		if (!this.validate(this.value)) {
			throw new InvalidValueObjectError('Invalid phone number format')
		}
	}

	protected validate(value: string): boolean {
		return PhoneNumberValueObject.PHONE_REGEX.test(value)
	}

	/**
	 * Formats the number with spaces
	 */
	toFormattedString(): string {
		return this.value.replace(/(\d{1,3})(?=(\d{3})+(?!\d))/g, '$1 ')
	}

	/**
	 * Gets the country calling code
	 */
	get countryCode(): string {
		return this.value.substring(0, this.value.length - 10)
	}
}
